crates/zkvm/adapters/sp1/src/prover.rs

```diff
@@ -28,44 +36,116 @@ impl ZKVMHost for SP1Host {
             prover_client: Arc::new(prover_client),
             proving_key,
             vkey,
+            elf: guest_code,
         }
     }
 
     fn prove<'a>(
         &self,
         prover_input: <Self::Input<'a> as ZKVMInputBuilder<'a>>::Input,
-    ) -> anyhow::Result<(Proof, VerificationKey)> {
+    ) -> anyhow::Result<(ProofWithMetadata, VerificationKey)> {
         // Init the prover
         if self.prover_options.use_mock_prover {
             std::env::set_var("SP1_PROVER", "mock");
             let mock_proof = Proof::new(vec![]);
+            let mock_proof_with_metadata =
+                ProofWithMetadata::new("mock_proof".to_owned(), mock_proof, None);
             let mock_vk = VerificationKey::new(vec![]);
-            return Ok((mock_proof, mock_vk));
+            return Ok((mock_proof_with_metadata, mock_vk));
         }
 
         let client = self.prover_client.clone();
 
+        // Generate unique ID for the proof
+        let mut input = bincode::serialize(&prover_input)?;
+        input.extend_from_slice(&self.vkey.hash_bytes());
+        let proof_id = format!("{}", strata_primitives::hash::raw(&input));
+
         // Start proving
-        let mut prover = client.prove(&self.proving_key, prover_input);
+        let mut prover = client.prove(&self.proving_key, prover_input.clone());
         if self.prover_options.enable_compression {
             prover = prover.compressed();
         }
         if self.prover_options.stark_to_snark_conversion {
             prover = prover.groth16();
         }
 
-        let proof = prover.run()?;
+        let (remote_id, proof_data) = if client.prover.id() == ProverType::Network {
+            println!("Remote prover");
+            let network_prover =
+                unsafe { &*(client.prover.as_ref() as *const _ as *const NetworkProver) };
+
+            let mode = match (
+                self.prover_options.enable_compression,
+                self.prover_options.stark_to_snark_conversion,
+            ) {
+                (_, true) => ProofMode::Groth16,
+                (true, _) => ProofMode::Compressed,
+                (_, _) => ProofMode::default(),
+            };
+            println!("Prover Mode: {:?}", mode);
+
+            let remote_id =
+                block_on(network_prover.request_proof(&self.elf, prover_input.clone(), mode))?;
+
+            let proof_data: SP1ProofWithPublicValues =
+                block_on(network_prover.wait_proof(&remote_id, None))?;
+
+            let filename: String = format!("{}.{}proof", remote_id, self.prover_options);
+            let mut file = File::create(filename).unwrap();
+            file.write_all(&bincode::serialize(&proof_data).unwrap())
+                .unwrap();
+
```

Comment:
I think this is done for the debugging purpose only right ?

Reply:
Yup.